ksql-engine/src/main/java/io/confluent/ksql/physical/PhysicalPlanBuilder.java

```diff
@@ -437,9 +443,24 @@ private static void throwIncompatibleKeysException(
   // Package private because of test
   String getServiceId() {
     return KsqlConstants.KSQL_INTERNAL_TOPIC_PREFIX
-           + ksqlConfig.getString(KsqlConfig.KSQL_SERVICE_ID_CONFIG);
+        + ksqlConfig.getString(KsqlConfig.KSQL_SERVICE_ID_CONFIG);
+  }
+
+  private String appendClientUsername(final String prefix, final ServiceContext serviceContext) {
+    final StringBuilder sb = new StringBuilder(prefix);
+
+    if (serviceContext.getContextType() == ServiceContext.ContextType.CLIENT_CONTEXT
+        && ksqlConfig.getBoolean(KsqlConfig.KSQL_TRANSIENT_APPLICATION_ID_APPEND_USERNAME)
+    ) {
+      serviceContext.getUsername().ifPresent(
+          user -> sb.append(Sanitizer.sanitize(user)).append("_")
```

Comment:
Doesn't this sanitizer allow through invalid characters, such as `%`? 
https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/common/utils/Sanitizer.java#L55

Reply:
Yes, I don't know how I missed this. I will need to create our own sanitizer to encode the % characters too. This might be a problem, though. For instance, I was thinking I could sanitize the name normally, and then replace the `%` character for another `_`. That would encode `org/user1` -> `org%2Fuser1`, then to `org_2Fuser1` which can be used for the name. But what if there is a prefix to allow users that start with `org_`?

I've been thinking on this approach of using the username in the topic, but I am starting to reject it. It might reduce security problems (as described in the PR), but not all of them. If we allow users, like a prefixed `user_`, to read from the table internal topic, and LDAP has users like `user_1`, `user+devs`, `user@confluent.com`, then all those users will start with `user_` (or whatever allowed chars we choose).

I'm thinking of another approach. I will explain it after all the comments below.